models/migrations/v58.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2017 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"fmt"
+
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+
+	"github.com/go-sql-driver/mysql"
+	"github.com/go-xorm/xorm"
+)
+
+func mysqlColumnsToUTF8MB4(x *xorm.Engine) (err error) {
+	if !setting.UseMySQL {
+		log.Info("Nothing to do")
+		return nil
+	}
+
+	const maxvc = 191
+	var migration_success = true
+
+	tables, err := x.DBMetas()
+	if err != nil {
+		return fmt.Errorf("cannot get tables: %v", err)
+	}
+	for _, table := range tables {
+		var ready_for_conversion = true
+		for ix := range table.Indexes {
+			if col := table.GetColumn(ix); col != nil {
+				if col.SQLType.Name == "VARCHAR" && col.Length > maxvc {
+					log.Info("reducing column %s.%s from %d to %d bytes", table.Name, ix, col.Length, maxvc)
+					var sqlstmt = fmt.Sprintf("alter table `%s` change column `%s` `%s` varchar(%d)", table.Name, ix, ix, maxvc)
+					if _, err := x.Exec(sqlstmt); err != nil {
+						if e, ok := err.(*mysql.MySQLError); ok {
+							if e.Number == 1265 || e.Number == 1406 {
+								log.Warn("failed. Please cut all data of this column down to a maximum of %d bytes", maxvc)
+							} else {
+								log.Warn("failed with %v", err)
+							}
+							ready_for_conversion = false
+							migration_success = false
+						} else {
+							return fmt.Errorf("something went horribly wrong: %v", err)
+						}
+					}
+				}
+			}
```

Comment:
This level of indentation makes me a bit dizzy :dizzy_face: 

```go
col := table.GetColumn(ix)
if col != nil {
  continue
}
if col.SQLType.Name != "VARCHAR" || col.Length <= maxvc {
  continue
}
```

and so on would be better I think. (Except for the error handling at the end. This is sort of my style to write Go code, so feel free to trash my opinion if you feel I'm wrong, but my general approach is to a) handle at the top level indentation the case we are looking for like varchar indexes, and b) handle in if branches the cases where something differs from the case we are looking for. Errors are often still unexpected and thus deserve to be placed in an if branch.)

Reply:
you're right. Such indentations are the result of grown code. Sorry for the eye cancer.